tests/unit/test_reduce_by_key.cu

```diff
@@ -0,0 +1,107 @@
+#include "../gtest.h"
+
+#include <vector>
+
+#include <backends/gpu/kernels/reduce_by_key.hpp>
+#include <memory/memory.hpp>
+#include <util/span.hpp>
+#include <util/rangeutil.hpp>
+
+using namespace nest::mc;
+
+template <typename T, typename I>
+__global__
+void reduce_kernel(const T* src, T* dst, const I* index, int n) {
+    unsigned tid = threadIdx.x + blockIdx.x*blockDim.x;
+
+    if (tid<n) {
+        gpu::reduce_by_key(src[tid], dst, index[tid]);
+    }
+}
+
+template <typename T>
+void reduce(const std::vector<T>& in, std::vector<T>& out, const std::vector<int>& index, unsigned block_dim=128) {
+    EXPECT_EQ(in.size(), index.size());
+
```

Comment:
Just to make things a bit more robust, would you consider testing for monotonicity of `index`, and automatically creating, zeroing and returning the `out` vector to suit? It would also remove a bit of the code in the test functions below.